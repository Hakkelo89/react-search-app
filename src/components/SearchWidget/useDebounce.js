import { useState, useEffect } from 'react'

const useDebounce = (searchTerm) => {
  const [value, setValue] = useState(searchTerm)

  useEffect(() => {
    const timer = setTimeout(() => {
      setValue(searchTerm)
    }, 500)

    return () => {
      clearTimeout(timer)
    }
  }, [searchTerm])
  
  return (value && value.length > 1) ? `https://www.rentalcars.com/FTSAutocomplete.do?solrIndex=fts_en&solrRows=10&solrTerm=${value}` : ''
}

export default useDebounce