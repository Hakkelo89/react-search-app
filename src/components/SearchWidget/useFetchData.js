import { useState, useEffect } from 'react'
import axios from 'axios'

const useFetchData = (searchURL) => {
  const [results, setResults] = useState([]) 
  
  useEffect(() => {
    if (searchURL) {
      const fetchData = async () => {
        const { data } = await axios.get(searchURL)
        const transformedData = data.results.docs.map(({ name }) => name)
        setResults(transformedData)
      }
      fetchData()
    } else setResults([])
  }, [searchURL])

  return results
}

export default useFetchData