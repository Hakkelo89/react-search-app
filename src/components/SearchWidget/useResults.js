import { useState } from 'react'

import useFetchData from './useFetchData'
import useDebounce from './useDebounce'

const useResults = (initialState = {}) => { 
  const {
    initialShowResult = false,
    initialSearchTerm = '',
  } = initialState
  const [showResults, setShowResults] = useState(initialShowResult)
  const [searchTerm, setSearchTerm] = useState(initialSearchTerm)

  const onBlur = () => setShowResults(false)

  const onFocus = () => setShowResults(true)

  const onChange = ({ target: { value } }) => setSearchTerm(value)

  const searchURL = useDebounce(searchTerm)
  
  const data = useFetchData(searchURL)

  const props = {
    onBlur,
    onFocus,
    onChange,
  }
  
  return [props, showResults, data]
}

export default useResults