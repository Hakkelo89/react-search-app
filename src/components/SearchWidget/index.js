import React from 'react'
import PropTypes from 'prop-types'

import useResults from './useResults'
import SearchResults from '../SearchResults'

const SearchWidget = ({ label, title, placeholder }) => {
  const [inputProps, showResults, results] = useResults()

  return (
    <div className="c-searchwidget">
      <h2 className="c-searchwidget__title">{title}</h2>
      <div className="c-searchwidget__searchbox">
        <label className="c-form-field__label">{label}</label>
        <div className="c-form-field__input">
          <input
            type="text"
            placeholder={placeholder}
            {...inputProps}
          />
          <SearchResults showResults={showResults} results={results} />
        </div>
      </div>
    </div>
  )
}

SearchWidget.defaultProps = {
  label: 'Pick-up Location',
  placeholder: 'city, airport, station, region, district...',
  title: 'Where are you going?',
}

SearchWidget.propTypes = {
  label: PropTypes.string.isRequired,
  placeholder: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}

export default SearchWidget