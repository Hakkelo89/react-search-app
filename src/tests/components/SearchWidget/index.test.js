import React from 'react'
import { mount } from 'enzyme'

import SearchWidget from '../../../components/SearchWidget'
import useDebounce from '../../../components/SearchWidget/useDebounce'
import useFetchData from '../../../components/SearchWidget/useFetchData'

jest.mock('../../../components/SearchWidget/useDebounce')
jest.mock('../../../components/SearchWidget/useFetchData')

const render = (props = {}) => mount(<SearchWidget {...props} />)

afterEach(() => {    
  jest.clearAllMocks()
})

describe('Snapshot Tests', () => {
  test('should render the snapshot of the SearchWidget component', () => {
    expect(render()).toMatchSnapshot();
  })
})

describe('Render Tests', () => {
  test('should render the SearchWidget container', () => {
    expect(render().find('.c-searchwidget').exists()).toEqual(true)
  })
  
  test('should render the SearchWidget title', () => {
    expect(render().find('.c-searchwidget__title').exists()).toEqual(true)
  })
  
  test('should render the searchbox container', () => {
    expect(render().find('.c-searchwidget__searchbox').exists()).toEqual(true)
  })
  
  test('should render the form field label', () => {
    expect(render().find('.c-form-field__label').exists()).toEqual(true)
  })
  
  test('should render the form-field input container', () => {
    expect(render().find('.c-form-field__input').exists()).toEqual(true)
  })
  
  test('should render an input element', () => {
    expect(render().find('input').exists()).toEqual(true)
  })
})

describe('SearchWidget Props Tests', () => {
  test('should render the SearchWidget title with default title', () => {
    expect(render().find('.c-searchwidget__title').text()).toEqual('Where are you going?')
  })

  test('should render the form field label with default label', () => {
    expect(render().find('.c-form-field__label').text()).toEqual('Pick-up Location')
  })

  test('should render an input element with default placeholder', () => {
    expect(render().find('input').prop('placeholder')).toEqual('city, airport, station, region, district...')
  })
  
  test('should render an input element with type \'text\'', () => {
    expect(render().find('input').prop('type')).toEqual('text')
  })

  test('should render the SearchWidget component with the expected title prop', () => {
    const title = 'Foo'
    expect(render({ title }).find('.c-searchwidget__title').text()).toEqual(title)
  })
  
  test('should render the SearchWidget component with the expected label prop', () => {
    const label = 'Foo'
    expect(render({ label }).find('.c-form-field__label').text()).toEqual(label)
  })
  
  test('should render the SearchWidget component with the expected placeholder prop', () => {
    const placeholder = 'Foo'
    expect(render({ placeholder }).find('input').prop('placeholder')).toEqual(placeholder)
  })
})

describe('SearchWidget Accessibility Test', () => {
  test('should make association between label and input element', () => {
    const wrapper = render()
    expect(wrapper.find('.c-form-field__label').prop('htmlFor')).toEqual('pick-up-location')
    expect(wrapper.find('input').prop('id')).toEqual('pick-up-location')
  })
  
})

describe('Search Results', () => {
  test('should not render on initial render', () => {
    expect(render().find('.c-search-results').exists()).toEqual(false)
  })

  test('should not render on onFocus event when results are empty', () => {
    const wrapper = render()
    const input = wrapper.find('input')
    input.simulate('focus')
  
    expect(wrapper.find('.c-search-results').exists()).toEqual(false)
  })

  test('should render on onFocus event when results are not empty', () => {
    const wrapper = render()
    const input = wrapper.find('input')

    useDebounce.mockReturnValue('someURL')
    useFetchData.mockReturnValue(['Foo', 'Bar'])

    input.simulate('change', { target: { value: 'Fo' } })
    input.simulate('focus')

    expect(wrapper.find('.c-search-results').exists()).toEqual(true)
  })

  test('should render on onChange event when length is greater than 1', () => {
    const wrapper = render()
    const input = wrapper.find('input')

    useDebounce.mockReturnValue('someURL')
    useFetchData.mockReturnValue(['Foo', 'Bar'])

    input.simulate('focus')
    input.simulate('change', { target: { value: 'Foo' } })
  
    expect(wrapper.find('.c-search-results').exists()).toEqual(true)
  })

  test('should not render on onChange event when length is equal to 1', () => {
    const wrapper = render()
    const input = wrapper.find('input')

    useDebounce.mockReturnValue('someURL')
    useFetchData.mockReturnValue([])

    input.simulate('focus')
    input.simulate('change', { target: { value: 'F' } })
  
    expect(wrapper.find('.c-search-results').exists()).toEqual(false)
  })

  test('should not render on onBlur event when results are not empty', () => {
    const wrapper = render()
    const input = wrapper.find('input')

    useDebounce.mockReturnValue('someURL')
    useFetchData.mockReturnValue(['Foo', 'Bar'])

    input.simulate('focus')
    input.simulate('change', { target: { value: 'Fo' } })
  
    expect(wrapper.find('.c-search-results').exists()).toEqual(true)

    input.simulate('blur')

    expect(wrapper.find('.c-search-results').exists()).toEqual(false)
  })
})