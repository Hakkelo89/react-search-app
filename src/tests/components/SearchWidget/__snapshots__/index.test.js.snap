// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot Tests should render the snapshot of the SearchWidget component 1`] = `
<SearchWidget
  label="Pick-up Location"
  placeholder="city, airport, station, region, district..."
  title="Where are you going?"
>
  <div
    className="c-searchwidget"
  >
    <h2
      className="c-searchwidget__title"
    >
      Where are you going?
    </h2>
    <div
      className="c-searchwidget__searchbox"
    >
      <label
        className="c-form-field__label"
      >
        Pick-up Location
      </label>
      <div
        className="c-form-field__input"
      >
        <input
          onBlur={[Function]}
          onChange={[Function]}
          onFocus={[Function]}
          placeholder="city, airport, station, region, district..."
          type="text"
        />
        <SearchResults
          results={Array []}
          showResults={false}
        />
      </div>
    </div>
  </div>
</SearchWidget>
`;
